packages/@aws-cdk/core/lib/asset-staging.ts

```diff
@@ -89,10 +89,15 @@ export class AssetStaging extends Construct {
     this.fingerprintOptions = props;
 
     if (props.bundling) {
-      this.bundleDir = this.bundle(props.bundling);
-    }
+      const assetHashIfKnownInAdvance = props.assetHashType !== AssetHashType.BUNDLE
+        ? this.calculateHash(props)
+        : undefined;
 
-    this.assetHash = this.calculateHash(props);
+      this.bundleDir = this.bundle(props.bundling, assetHashIfKnownInAdvance);
+      this.assetHash = assetHashIfKnownInAdvance ?? this.calculateHash(props);
```

Comment:
Not sure I like the fact the calculateHash accepts `props` can we just pass the relevant once for readability?

Reply:
I've refactored calculateHash to accept a hashType and custom assetHash parameter which is used when the hashType is CUSTOM.